Old records should be removed from PEERS table after node ip address was changed

*Installation details*
Scylla version (or git commit hash): 5.5.0~dev-0.20240105.7e84e03f5231 with build-id f21e4548b69223a75d01fd3bb9d4c9c2b1b71a6d
Cluster size: 3

Dtest `update_cluster_layout_tests.TestUpdateClusterLayout.test_change_node_ip` failed with error when experimental feature:`consistent-topology-changes` is enabled.
```
>           assert len(peers) == 2, f"There are more peers than expected.{peers}"
E           AssertionError: There are more peers than expected.[[UUID('e59eda2f-fffb-484c-b14b-8bb46e20ddc2'), '127.0.96.33'], [UUID('bf05eab8-e210-4d32-ba12-f95e1c168bf3'), '127.0.96.2'], [UUID('e59eda2f-fffb-484c-b14b-8bb46e20ddc2'), '127.0.96.3']]
E           assert 3 == 2
E            +  where 3 = len([[UUID('e59eda2f-fffb-484c-b14b-8bb46e20ddc2'), '127.0.96.33'], [UUID('bf05eab8-e210-4d32-ba12-f95e1c168bf3'), '127.0.96.2'], [UUID('e59eda2f-fffb-484c-b14b-8bb46e20ddc2'), '127.0.96.3']])

update_cluster_layout_tests.py:2256: AssertionError

```
this happened, because PEER table on node1 and node2 contains records with old and new ip for same host_id for node3, after node3 was started with new ip address.

run dtest:
`pytest --log-cli-level debug --keep-test-dir --scylla-version=unstable/master:latest --experimental-features=consistent-topology-changes  update_cluster_layout_tests.py::TestUpdateClusterLayout::test_change_node_ip`

test debug logs:
```
17:21:56,812 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2297 | Change IP address for node3
17:21:56,836 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2302 | Start node3 again with ip address 127.0.54.33
17:21:56,840 ccm             DEBUG cluster.py     :754  | node3: Starting scylla: args=['/home/abykov/.dtest/dtest-5k72mpz9/test/node3/bin/scylla', '--options-file', '/home/abykov/.dtest/dtest-5k72mpz9/test/node3/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.54.33', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--overprovisioned', '--prometheus-address', '127.0.54.33', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=False wait_for_binary_proto=True
17:21:57,120 cassandra.pool  WARNING thread.py      :58   | Error attempting to reconnect to 127.0.54.3:9042, scheduling retry in 4.0 seconds: [Errno 111] Tried connecting to [('127.0.54.3', 9042)]. Last error: Connection refused
17:21:57,180 ccm             DEBUG cluster.py     :754  | node3: Starting scylla-jmx: args=['/home/abykov/.dtest/dtest-5k72mpz9/test/node3/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.54.33', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.54.33', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.54.33', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/home/abykov/.dtest/dtest-5k72mpz9/test/node3/bin/scylla-jmx-1.0.jar']
17:21:57,532 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2305 | Node3 is now up
17:21:57,533 urllib3.connectionpool DEBUG connectionpool.py:246  | Starting new HTTP connection (1): 127.0.54.33:10000
17:21:57,534 urllib3.connectionpool DEBUG connectionpool.py:474  | http://127.0.54.33:10000 "GET /storage_service/hostid/local HTTP/1.1" 200 38
17:21:57,535 ccm             DEBUG cluster.py     :754  | node1: nodetool cmd=status  wait=True timeout=None
17:21:58,639 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2310 | nodetool status from node1: {'Datacenter': 'datacenter1', 'nodes': [{'status': 'UN', 'address': '127.0.54.1', 'load': '211.84 KB', 'tokens': '256', 'owns': '?', 'host id': '35b7d7b0-f8be-49f2-a257-bf36b7749992', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.2', 'load': '448.37 KB', 'tokens': '256', 'owns': '?', 'host id': 'ac138eb5-8b27-430a-b67a-1008fa61876a', 'rack': 'rack1'}, {'status': 'DN', 'address': '127.0.54.3', 'load': '480.51 KB', 'tokens': '256', 'owns': '?', 'host id': 'ee1ade88-3304-4366-a83e-f15e2e3db31f', 'rack': 'rack1'}]}
17:21:58,641 ccm             DEBUG cluster.py     :754  | node2: nodetool cmd=status  wait=True timeout=None
17:21:59,866 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2310 | nodetool status from node2: {'Datacenter': 'datacenter1', 'nodes': [{'status': 'UN', 'address': '127.0.54.33', 'load': '?', 'tokens': '256', 'owns': '?', 'host id': 'ee1ade88-3304-4366-a83e-f15e2e3db31f', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.1', 'load': '?', 'tokens': '256', 'owns': '?', 'host id': '35b7d7b0-f8be-49f2-a257-bf36b7749992', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.2', 'load': '448.37 KB', 'tokens': '256', 'owns': '?', 'host id': 'ac138eb5-8b27-430a-b67a-1008fa61876a', 'rack': 'rack1'}]}
17:21:59,867 ccm             DEBUG cluster.py     :754  | node3: nodetool cmd=status  wait=True timeout=None
17:22:01,019 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2310 | nodetool status from node3: {'Datacenter': 'datacenter1', 'nodes': [{'status': 'UN', 'address': '127.0.54.1', 'load': '211.84 KB', 'tokens': '256', 'owns': '?', 'host id': '35b7d7b0-f8be-49f2-a257-bf36b7749992', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.2', 'load': '448.37 KB', 'tokens': '256', 'owns': '?', 'host id': 'ac138eb5-8b27-430a-b67a-1008fa61876a', 'rack': 'rack1'}, {'status': 'DN', 'address': '127.0.54.3', 'load': '480.51 KB', 'tokens': '256', 'owns': '?', 'host id': 'ee1ade88-3304-4366-a83e-f15e2e3db31f', 'rack': 'rack1'}]}
17:22:01,020 ccm             DEBUG cluster.py     :754  | node1: nodetool cmd=status  wait=True timeout=None
17:22:01,126 cassandra.pool  WARNING thread.py      :58   | Error attempting to reconnect to 127.0.54.3:9042, scheduling retry in 4.0 seconds: [Errno 111] Tried connecting to [('127.0.54.3', 9042)]. Last error: Connection refused
17:22:02,032 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2249 | nodetool status from node1: {'Datacenter': 'datacenter1', 'nodes': [{'status': 'UN', 'address': '127.0.54.33', 'load': '1.47 MB', 'tokens': '256', 'owns': '?', 'host id': 'ee1ade88-3304-4366-a83e-f15e2e3db31f', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.1', 'load': '211.84 KB', 'tokens': '256', 'owns': '?', 'host id': '35b7d7b0-f8be-49f2-a257-bf36b7749992', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.2', 'load': '448.37 KB', 'tokens': '256', 'owns': '?', 'host id': 'ac138eb5-8b27-430a-b67a-1008fa61876a', 'rack': 'rack1'}]}
17:22:02,037 cassandra.cluster INFO  dtest_setup.py :437  | New Cassandra host <Host: 127.0.54.3:9042 datacenter1> discovered
17:22:02,037 cassandra.cluster WARNING dtest_setup.py :437  | Found multiple hosts with the same host_id (ee1ade88-3304-4366-a83e-f15e2e3db31f). Excluding peer 127.0.54.33
17:22:02,037 cassandra.cluster INFO  dtest_setup.py :437  | New Cassandra host <Host: 127.0.54.2:9042 datacenter1> discovered
17:22:02,052 update_cluster_layout_tests INFO  update_cluster_layout_tests.py:2252 | PEERS: [[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2']]
17:22:02,053 update_cluster_layout_tests INFO  update_cluster_layout_tests.py:2254 | LOCAL: [[UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1']]
17:22:02,053 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2258 | Check peer table for node1 with ip address 127.0.54.1 : peers=[[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2']] 
17:22:02,054 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2259 | Check peer table for node1 with ip address 127.0.54.1 : local=[[UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1']] 
17:22:02,054 ccm             DEBUG cluster.py     :754  | node2: nodetool cmd=status  wait=True timeout=None
17:22:03,069 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2249 | nodetool status from node2: {'Datacenter': 'datacenter1', 'nodes': [{'status': 'UN', 'address': '127.0.54.33', 'load': '1.47 MB', 'tokens': '256', 'owns': '?', 'host id': 'ee1ade88-3304-4366-a83e-f15e2e3db31f', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.1', 'load': '?', 'tokens': '256', 'owns': '?', 'host id': '35b7d7b0-f8be-49f2-a257-bf36b7749992', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.2', 'load': '448.37 KB', 'tokens': '256', 'owns': '?', 'host id': 'ac138eb5-8b27-430a-b67a-1008fa61876a', 'rack': 'rack1'}]}
17:22:03,073 cassandra.cluster INFO  dtest_setup.py :437  | New Cassandra host <Host: 127.0.54.3:9042 datacenter1> discovered
17:22:03,074 cassandra.cluster WARNING dtest_setup.py :437  | Found multiple hosts with the same host_id (ee1ade88-3304-4366-a83e-f15e2e3db31f). Excluding peer 127.0.54.33
17:22:03,074 cassandra.cluster INFO  dtest_setup.py :437  | New Cassandra host <Host: 127.0.54.1:9042 datacenter1> discovered
17:22:03,093 update_cluster_layout_tests INFO  update_cluster_layout_tests.py:2252 | PEERS: [[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33'], [UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1']]
17:22:03,094 update_cluster_layout_tests INFO  update_cluster_layout_tests.py:2254 | LOCAL: [[UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2']]
17:22:03,095 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2258 | Check peer table for node2 with ip address 127.0.54.2 : peers=[[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33'], [UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1']] 
17:22:03,095 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2259 | Check peer table for node2 with ip address 127.0.54.2 : local=[[UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2']] 
17:22:03,096 ccm             DEBUG cluster.py     :754  | node3: nodetool cmd=status  wait=True timeout=None
17:22:03,984 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2249 | nodetool status from node3: {'Datacenter': 'datacenter1', 'nodes': [{'status': 'UN', 'address': '127.0.54.1', 'load': '211.84 KB', 'tokens': '256', 'owns': '?', 'host id': '35b7d7b0-f8be-49f2-a257-bf36b7749992', 'rack': 'rack1'}, {'status': 'UN', 'address': '127.0.54.2', 'load': '448.37 KB', 'tokens': '256', 'owns': '?', 'host id': 'ac138eb5-8b27-430a-b67a-1008fa61876a', 'rack': 'rack1'}, {'status': 'DN', 'address': '127.0.54.3', 'load': '480.51 KB', 'tokens': '256', 'owns': '?', 'host id': 'ee1ade88-3304-4366-a83e-f15e2e3db31f', 'rack': 'rack1'}]}
17:22:03,988 cassandra.cluster WARNING dtest_setup.py :437  | Found multiple hosts with the same host_id (ee1ade88-3304-4366-a83e-f15e2e3db31f). Excluding peer 127.0.54.3
17:22:03,988 cassandra.cluster INFO  dtest_setup.py :437  | New Cassandra host <Host: 127.0.54.2:9042 datacenter1> discovered
17:22:03,989 cassandra.cluster INFO  dtest_setup.py :437  | New Cassandra host <Host: 127.0.54.1:9042 datacenter1> discovered
17:22:04,004 update_cluster_layout_tests INFO  update_cluster_layout_tests.py:2252 | PEERS: [[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2'], [UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1']]
17:22:04,006 update_cluster_layout_tests INFO  update_cluster_layout_tests.py:2254 | LOCAL: [[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33']]
17:22:04,006 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2258 | Check peer table for node3 with ip address 127.0.54.33 : peers=[[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2'], [UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1']] 
17:22:04,006 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2259 | Check peer table for node3 with ip address 127.0.54.33 : local=[[UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33']] 
17:22:04,007 update_cluster_layout_tests DEBUG update_cluster_layout_tests.py:2262 | peers_and_local=["[[UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33']]", "[[UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33']]", "[[UUID('35b7d7b0-f8be-49f2-a257-bf36b7749992'), '127.0.54.1'], [UUID('ac138eb5-8b27-430a-b67a-1008fa61876a'), '127.0.54.2'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.3'], [UUID('ee1ade88-3304-4366-a83e-f15e2e3db31f'), '127.0.54.33']]"]
```

The issue should be fixed by: https://github.com/scylladb/scylladb/pull/16639

--- a/db/system_keyspace.cc
+++ b/db/system_keyspace.cc
@@ -1874,6 +1874,12 @@ static data_value_or_unset make_data_value_or_unset(const std::optional<std::uno
 }
 
 future<> system_keyspace::update_peer_info(gms::inet_address ep, locator::host_id hid, const peer_info& info) {
+    if (ep == gms::inet_address{}) {
+        on_internal_error(slogger, format("update_peer_info called with empty inet_address, host_id {}", hid));
+    }
+    if (!hid) {
+        on_internal_error(slogger, format("update_peer_info called with empty host_id, ep {}", ep));
+    }
     if (_db.get_token_metadata().get_topology().is_me(ep)) {
         on_internal_error(slogger, format("update_peer_info called for this node: {}", ep));
     }

--- a/db/system_keyspace.cc
+++ b/db/system_keyspace.cc
@@ -1465,19 +1465,17 @@ struct local_cache {
 };
 
 future<std::unordered_map<gms::inet_address, locator::endpoint_dc_rack>> system_keyspace::load_dc_rack_info() {
-    auto msg = co_await execute_cql(format("SELECT peer, data_center, rack from system.{}", PEERS));
+    const auto msg = co_await execute_cql(format("SELECT peer, data_center, rack from system.{}", PEERS));
 
     std::unordered_map<gms::inet_address, locator::endpoint_dc_rack> ret;
     for (const auto& row : *msg) {
-        net::inet_address peer = row.template get_as<net::inet_address>("peer");
         if (!row.has("data_center") || !row.has("rack")) {
             continue;
         }
-        gms::inet_address gms_addr(std::move(peer));
-        sstring dc = row.template get_as<sstring>("data_center");
-        sstring rack = row.template get_as<sstring>("rack");
-
-        ret.emplace(gms_addr, locator::endpoint_dc_rack{ dc, rack });
+        ret.emplace(row.get_as<net::inet_address>("peer"), locator::endpoint_dc_rack {
+            row.get_as<sstring>("data_center"),
+            row.get_as<sstring>("rack")
+        });
     }
 
     co_return ret;
@@ -1747,43 +1745,41 @@ static locator::host_id get_host_id(const gms::inet_address& peer, const cql3::u
 }
 
 future<std::unordered_map<gms::inet_address, std::unordered_set<dht::token>>> system_keyspace::load_tokens() {
-    sstring req = format("SELECT peer, host_id, tokens FROM system.{}", PEERS);
-    return execute_cql(req).then([] (::shared_ptr<cql3::untyped_result_set> cql_result) {
-        std::unordered_map<gms::inet_address, std::unordered_set<dht::token>> ret;
-        for (auto& row : *cql_result) {
-            auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
-            if (!get_host_id(peer, row)) {
-                continue;
-            }
-            if (row.has("tokens")) {
-                ret.emplace(peer, decode_tokens(deserialize_set_column(*peers(), row, "tokens")));
-            }
+    const sstring req = format("SELECT peer, host_id, tokens FROM system.{}", PEERS);
+    std::unordered_map<gms::inet_address, std::unordered_set<dht::token>> ret;
+    const auto cql_result = co_await execute_cql(req);
+    for (const auto& row : *cql_result) {
+        const auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
+        if (!get_host_id(peer, row)) {
+            continue;
         }
-        return ret;
-    });
+        if (row.has("tokens")) {
+            ret.emplace(peer, decode_tokens(deserialize_set_column(*peers(), row, "tokens")));
+        }
+    }
+    co_return ret;
 }
 
 future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace::load_host_ids() {
-    sstring req = format("SELECT peer, host_id FROM system.{}", PEERS);
-    return execute_cql(req).then([] (::shared_ptr<cql3::untyped_result_set> cql_result) {
-        std::unordered_map<gms::inet_address, locator::host_id> ret;
-        for (auto& row : *cql_result) {
-            auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
-            if (auto host_id = get_host_id(peer, row)) {
-                ret.emplace(peer, host_id);
-            }
+    const sstring req = format("SELECT peer, host_id FROM system.{}", PEERS);
+    std::unordered_map<gms::inet_address, locator::host_id> ret;
+    const auto cql_result = co_await execute_cql(req);
+    for (const auto& row : *cql_result) {
+        const auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
+        if (auto host_id = get_host_id(peer, row)) {
+            ret.emplace(peer, host_id);
         }
-        return ret;
-    });
+    }
+    co_return ret;
 }
 
 future<std::vector<gms::inet_address>> system_keyspace::load_peers() {
-    auto res = co_await execute_cql(format("SELECT peer, host_id, tokens FROM system.{}", PEERS));
+    const auto res = co_await execute_cql(format("SELECT peer, host_id, tokens FROM system.{}", PEERS));
     assert(res);
 
     std::vector<gms::inet_address> ret;
-    for (auto& row: *res) {
-        auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
+    for (const auto& row: *res) {
+        const auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
         if (!get_host_id(peer, row)) {
             continue;
         }
@@ -1800,33 +1796,31 @@ future<std::vector<gms::inet_address>> system_keyspace::load_peers() {
 }
 
 future<std::unordered_map<gms::inet_address, sstring>> system_keyspace::load_peer_features() {
-    sstring req = format("SELECT peer, supported_features FROM system.{}", PEERS);
-    return execute_cql(req).then([] (::shared_ptr<cql3::untyped_result_set> cql_result) {
-        std::unordered_map<gms::inet_address, sstring> ret;
-        for (auto& row : *cql_result) {
-            if (row.has("supported_features")) {
-                ret.emplace(row.get_as<net::inet_address>("peer"),
-                        row.get_as<sstring>("supported_features"));
-            }
+    const sstring req = format("SELECT peer, supported_features FROM system.{}", PEERS);
+    std::unordered_map<gms::inet_address, sstring> ret;
+    const auto cql_result = co_await execute_cql(req);
+    for (const auto& row : *cql_result) {
+        if (row.has("supported_features")) {
+            ret.emplace(row.get_as<net::inet_address>("peer"),
+                    row.get_as<sstring>("supported_features"));
         }
-        return ret;
-    });
+    }
+    co_return ret;
 }
 
 future<std::unordered_map<gms::inet_address, gms::inet_address>> system_keyspace::get_preferred_ips() {
-    sstring req = format("SELECT peer, preferred_ip FROM system.{}", PEERS);
-    return execute_cql(req).then([] (::shared_ptr<cql3::untyped_result_set> cql_res_set) {
-        std::unordered_map<gms::inet_address, gms::inet_address> res;
+    const sstring req = format("SELECT peer, preferred_ip FROM system.{}", PEERS);
+    std::unordered_map<gms::inet_address, gms::inet_address> res;
 
-        for (auto& r : *cql_res_set) {
-            if (r.has("preferred_ip")) {
-                res.emplace(gms::inet_address(r.get_as<net::inet_address>("peer")),
-                            gms::inet_address(r.get_as<net::inet_address>("preferred_ip")));
-            }
+    const auto cql_result = co_await execute_cql(req);
+    for (const auto& r : *cql_result) {
+        if (r.has("preferred_ip")) {
+            res.emplace(gms::inet_address(r.get_as<net::inet_address>("peer")),
+                        gms::inet_address(r.get_as<net::inet_address>("preferred_ip")));
         }
+    }
 
-        return res;
-    });
+    co_return res;
 }
 
 namespace {
@@ -1928,7 +1922,7 @@ future<> system_keyspace::update_schema_version(table_schema_version version) {
  * Remove stored tokens being used by another node
  */
 future<> system_keyspace::remove_endpoint(gms::inet_address ep) {
-    sstring req = format("DELETE FROM system.{} WHERE peer = ?", PEERS);
+    const sstring req = format("DELETE FROM system.{} WHERE peer = ?", PEERS);
     slogger.debug("DELETE FROM system.{} WHERE peer = {}", PEERS, ep);
     co_await execute_cql(req, ep.addr()).discard_result();
 }

--- a/db/system_keyspace.cc
+++ b/db/system_keyspace.cc
@@ -1464,7 +1464,47 @@ struct local_cache {
     system_keyspace::bootstrap_state _state;
 };
 
+future<> system_keyspace::peers_table_read_fixup() {
+    assert(this_shard_id() == 0);
+    if (_peers_table_read_fixup_done) {
+        co_return;
+    }
+    _peers_table_read_fixup_done = true;
+
+    const auto cql = format("SELECT peer, host_id, WRITETIME(host_id) as ts from system.{}", PEERS);
+    std::unordered_map<utils::UUID, std::pair<net::inet_address, int64_t>> map{};
+    const auto cql_result = co_await execute_cql(cql);
+    for (const auto& row : *cql_result) {
+        const auto peer = row.get_as<net::inet_address>("peer");
+        if (!row.has("host_id")) {
+            slogger.error("Peer {} has no host_id in system.{}, the record is broken, removing it",
+                peer, system_keyspace::PEERS);
+            co_await remove_endpoint(gms::inet_address{peer});
+            continue;
+        }
+        const auto host_id = row.get_as<utils::UUID>("host_id");
+        const auto ts = row.get_as<int64_t>("ts");
+        const auto it = map.find(host_id);
+        if (it == map.end()) {
+            map.insert({host_id, {peer, ts}});
+            continue;
+        }
+        if (it->second.second >= ts) {
+            slogger.error("Peer {} with host_id {} has newer IP {} in system.{}, the record is stale, removing it",
+                peer, host_id, it->second.first, system_keyspace::PEERS);
+            co_await remove_endpoint(gms::inet_address{peer});
+        } else {
+            slogger.error("Peer {} with host_id {} has newer IP {} in system.{}, the record is stale, removing it",
+                it->second.first, host_id, peer, system_keyspace::PEERS);
+            co_await remove_endpoint(gms::inet_address{it->second.first});
+            it->second = {peer, ts};
+        }
+    }
+}
+
 future<std::unordered_map<gms::inet_address, locator::endpoint_dc_rack>> system_keyspace::load_dc_rack_info() {
+    co_await peers_table_read_fixup();
+
     const auto msg = co_await execute_cql(format("SELECT peer, data_center, rack from system.{}", PEERS));
 
     std::unordered_map<gms::inet_address, locator::endpoint_dc_rack> ret;
@@ -1733,56 +1773,42 @@ static std::vector<cdc::generation_id_v2> decode_cdc_generations_ids(const set_t
     return gen_ids_list;
 }
 
-static locator::host_id get_host_id(const gms::inet_address& peer, const cql3::untyped_result_set::row& row) {
-    locator::host_id host_id;
-    if (row.has("host_id")) {
-        host_id = locator::host_id(row.get_as<utils::UUID>("host_id"));
-    }
-    if (!host_id) {
-        slogger.warn("Peer {} has no host_id in system.{}", peer, system_keyspace::PEERS);
-    }
-    return host_id;
-}
-
 future<std::unordered_map<gms::inet_address, std::unordered_set<dht::token>>> system_keyspace::load_tokens() {
-    const sstring req = format("SELECT peer, host_id, tokens FROM system.{}", PEERS);
+    co_await peers_table_read_fixup();
+
+    const sstring req = format("SELECT peer, tokens FROM system.{}", PEERS);
     std::unordered_map<gms::inet_address, std::unordered_set<dht::token>> ret;
     const auto cql_result = co_await execute_cql(req);
     for (const auto& row : *cql_result) {
-        const auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
-        if (!get_host_id(peer, row)) {
-            continue;
-        }
         if (row.has("tokens")) {
-            ret.emplace(peer, decode_tokens(deserialize_set_column(*peers(), row, "tokens")));
+            ret.emplace(gms::inet_address(row.get_as<net::inet_address>("peer")),
+                decode_tokens(deserialize_set_column(*peers(), row, "tokens")));
         }
     }
     co_return ret;
 }
 
 future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace::load_host_ids() {
+    co_await peers_table_read_fixup();
+
     const sstring req = format("SELECT peer, host_id FROM system.{}", PEERS);
     std::unordered_map<gms::inet_address, locator::host_id> ret;
     const auto cql_result = co_await execute_cql(req);
     for (const auto& row : *cql_result) {
-        const auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
-        if (auto host_id = get_host_id(peer, row)) {
-            ret.emplace(peer, host_id);
-        }
+        ret.emplace(gms::inet_address(row.get_as<net::inet_address>("peer")),
+            locator::host_id(row.get_as<utils::UUID>("host_id")));
     }
     co_return ret;
 }
 
 future<std::vector<gms::inet_address>> system_keyspace::load_peers() {
-    const auto res = co_await execute_cql(format("SELECT peer, host_id, tokens FROM system.{}", PEERS));
+    co_await peers_table_read_fixup();
+
+    const auto res = co_await execute_cql(format("SELECT peer, tokens FROM system.{}", PEERS));
     assert(res);
 
     std::vector<gms::inet_address> ret;
     for (const auto& row: *res) {
-        const auto peer = gms::inet_address(row.get_as<net::inet_address>("peer"));
-        if (!get_host_id(peer, row)) {
-            continue;
-        }
         if (!row.has("tokens")) {
             // Ignore rows that don't have tokens. Such rows may
             // be introduced by code that persists parts of peer
@@ -1790,12 +1816,14 @@ future<std::vector<gms::inet_address>> system_keyspace::load_peers() {
             // race with deleting a peer (during node removal).
             continue;
         }
-        ret.emplace_back(peer);
+        ret.emplace_back(gms::inet_address(row.get_as<net::inet_address>("peer")));
     }
     co_return ret;
 }
 
 future<std::unordered_map<gms::inet_address, sstring>> system_keyspace::load_peer_features() {
+    co_await peers_table_read_fixup();
+
     const sstring req = format("SELECT peer, supported_features FROM system.{}", PEERS);
     std::unordered_map<gms::inet_address, sstring> ret;
     const auto cql_result = co_await execute_cql(req);
@@ -1809,6 +1837,8 @@ future<std::unordered_map<gms::inet_address, sstring>> system_keyspace::load_pee
 }
 
 future<std::unordered_map<gms::inet_address, gms::inet_address>> system_keyspace::get_preferred_ips() {
+    co_await peers_table_read_fixup();
+
     const sstring req = format("SELECT peer, preferred_ip FROM system.{}", PEERS);
     std::unordered_map<gms::inet_address, gms::inet_address> res;
 

--- a/db/system_keyspace.cc
+++ b/db/system_keyspace.cc
@@ -1873,7 +1873,7 @@ static data_value_or_unset make_data_value_or_unset(const std::optional<std::uno
 };
 }
 
-future<> system_keyspace::update_peer_info(gms::inet_address ep, const peer_info& info) {
+future<> system_keyspace::update_peer_info(gms::inet_address ep, locator::host_id hid, const peer_info& info) {
     if (_db.get_token_metadata().get_topology().is_me(ep)) {
         on_internal_error(slogger, format("update_peer_info called for this node: {}", ep));
     }
@@ -1881,7 +1881,7 @@ future<> system_keyspace::update_peer_info(gms::inet_address ep, const peer_info
     data_value_list values = {
         data_value_or_unset(data_value(ep.addr())),
         make_data_value_or_unset(info.data_center),
-        make_data_value_or_unset(info.host_id),
+        data_value_or_unset(hid.id),
         make_data_value_or_unset(info.preferred_ip),
         make_data_value_or_unset(info.rack),
         make_data_value_or_unset(info.release_version),

--- a/db/system_keyspace.cc
+++ b/db/system_keyspace.cc
@@ -1855,18 +1855,16 @@ future<std::unordered_map<gms::inet_address, gms::inet_address>> system_keyspace
 
 namespace {
 template <typename T>
-static data_value_or_unset make_data_value_or_unset(const std::optional<T>& opt, bool& any_set) {
+static data_value_or_unset make_data_value_or_unset(const std::optional<T>& opt) {
     if (opt) {
-        any_set = true;
         return data_value(*opt);
     } else {
         return unset_value{};
     }
 };
 
-static data_value_or_unset make_data_value_or_unset(const std::optional<std::unordered_set<dht::token>>& opt, bool& any_set) {
+static data_value_or_unset make_data_value_or_unset(const std::optional<std::unordered_set<dht::token>>& opt) {
     if (opt) {
-        any_set = true;
         auto set_type = set_type_impl::get_instance(utf8_type, true);
         return make_set_value(set_type, prepare_tokens(*opt));
     } else {
@@ -1880,23 +1878,18 @@ future<> system_keyspace::update_peer_info(gms::inet_address ep, const peer_info
         on_internal_error(slogger, format("update_peer_info called for this node: {}", ep));
     }
 
-    bool any_set = false;
     data_value_list values = {
         data_value_or_unset(data_value(ep.addr())),
-        make_data_value_or_unset(info.data_center, any_set),
-        make_data_value_or_unset(info.host_id, any_set),
-        make_data_value_or_unset(info.preferred_ip, any_set),
-        make_data_value_or_unset(info.rack, any_set),
-        make_data_value_or_unset(info.release_version, any_set),
-        make_data_value_or_unset(info.rpc_address, any_set),
-        make_data_value_or_unset(info.schema_version, any_set),
-        make_data_value_or_unset(info.tokens, any_set),
-        make_data_value_or_unset(info.supported_features, any_set),
+        make_data_value_or_unset(info.data_center),
+        make_data_value_or_unset(info.host_id),
+        make_data_value_or_unset(info.preferred_ip),
+        make_data_value_or_unset(info.rack),
+        make_data_value_or_unset(info.release_version),
+        make_data_value_or_unset(info.rpc_address),
+        make_data_value_or_unset(info.schema_version),
+        make_data_value_or_unset(info.tokens),
+        make_data_value_or_unset(info.supported_features),
     };
-
-    if (!any_set) {
-        co_return;
-    }
 
     auto query = fmt::format("INSERT INTO system.{} "
             "(peer,data_center,host_id,preferred_ip,rack,release_version,rpc_address,schema_version,tokens,supported_features) VALUES"

--- a/db/system_keyspace.hh
+++ b/db/system_keyspace.hh
@@ -275,7 +275,6 @@ public:
 public:
     struct peer_info {
         std::optional<sstring> data_center;
-        std::optional<utils::UUID> host_id;
         std::optional<net::inet_address> preferred_ip;
         std::optional<sstring> rack;
         std::optional<sstring> release_version;
@@ -285,7 +284,7 @@ public:
         std::optional<sstring> supported_features;
     };
 
-    future<> update_peer_info(gms::inet_address ep, const peer_info& info);
+    future<> update_peer_info(gms::inet_address ep, locator::host_id hid, const peer_info& info);
 
     future<> remove_endpoint(gms::inet_address ep);
 

--- a/db/system_keyspace.hh
+++ b/db/system_keyspace.hh
@@ -114,6 +114,7 @@ class system_keyspace : public seastar::peering_sharded_service<system_keyspace>
     replica::database& _db;
     std::unique_ptr<local_cache> _cache;
     virtual_tables_registry _virtual_tables_registry;
+    bool _peers_table_read_fixup_done = false;
 
     static schema_ptr raft_snapshot_config();
     static schema_ptr local();
@@ -128,6 +129,15 @@ class system_keyspace : public seastar::peering_sharded_service<system_keyspace>
     static schema_ptr large_cells();
     static schema_ptr scylla_local();
     future<> force_blocking_flush(sstring cfname);
+    // This function is called when the system.peers table is read,
+    // and it fixes some types of inconsistencies that can occur
+    // due to node crashes:
+    //  * missing host_id. This is possible in the old versions of the code. Such records
+    //  are removed and the warning is written to the log.
+    //  * duplicate IPs for a given host_id. This is possible when some node changes its IP
+    //  and this node crashes after adding a new IP but before removing the old one. The
+    //  record with older timestamp is removed, the warning is written to the log.
+    future<> peers_table_read_fixup();
 public:
     static schema_ptr size_estimates();
 public:

--- a/main.cc
+++ b/main.cc
@@ -1792,8 +1792,10 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // Set up group0 service earlier since it is needed by group0 setup just below
             ss.local().set_group0(group0_service, raft_topology_change_enabled);
 
+            const auto generation_number = gms::generation_type(sys_ks.local().increment_and_get_generation().get());
+
             // Load address_map from system.peers and subscribe to gossiper events to keep it updated.
-            ss.local().init_address_map(raft_address_map.local()).get();
+            ss.local().init_address_map(raft_address_map.local(), generation_number).get();
             auto cancel_address_map_subscription = defer_verbose_shutdown("storage service uninit address map", [&ss] {
                 ss.local().uninit_address_map().get();
             });
@@ -1814,7 +1816,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             }).get();
 
             with_scheduling_group(maintenance_scheduling_group, [&] {
-                return ss.local().join_cluster(sys_dist_ks, proxy, gossiper, service::start_hint_manager::yes);
+                return ss.local().join_cluster(sys_dist_ks, proxy, gossiper, service::start_hint_manager::yes, generation_number);
             }).get();
 
             sl_controller.invoke_on_all([&lifecycle_notifier] (qos::service_level_controller& controller) {

--- a/service/raft/raft_group0.cc
+++ b/service/raft/raft_group0.cc
@@ -386,12 +386,6 @@ future<> raft_group0::start_server_for_group0(raft::group_id group0_id, service:
     // The address map may miss our own id in case we connect
     // to an existing Raft Group 0 leader.
     auto my_id = load_my_id();
-    _raft_gr.address_map().add_or_update_entry(my_id, _gossiper.get_broadcast_address());
-    // At this time the group registry is already up and running,
-    // so the address map is getting all the notifications from
-    // the gossiper. By reading the application state *after* subscribing to new gossip events,
-    // we ensure we haven't missed any IP update in the map.
-    load_initial_raft_address_map();
     group0_log.info("Server {} is starting group 0 with id {}", my_id, group0_id);
     auto srv_for_group0 = create_server_for_group0(group0_id, my_id, ss, qp, mm, topology_change_enabled);
     auto& persistence = srv_for_group0.persistence;
@@ -746,25 +740,6 @@ future<> raft_group0::setup_group0(
 
     group0_log.info("setup_group0: the cluster is ready to use Raft. Finishing.");
     co_await _client.set_group0_upgrade_state(group0_upgrade_state::use_post_raft_procedures);
-}
-
-void raft_group0::load_initial_raft_address_map() {
-    _gossiper.for_each_endpoint_state([this] (const gms::inet_address& ip_addr, const gms::endpoint_state& state) {
-        auto* value = state.get_application_state_ptr(gms::application_state::HOST_ID);
-        if (value == nullptr) {
-            return;
-        }
-        auto server_id = utils::UUID(value->value());
-        if (server_id == utils::UUID{}) {
-            upgrade_log.error("empty Host ID for host {} ", ip_addr);
-            return;
-        }
-        // The failure detector needs the IPs on all shards. We
-        // can safely overwrite existing entries since are loading
-        // them directly from gossiper app state - which is most
-        // recent.
-        _raft_gr.address_map().add_or_update_entry(raft::server_id{server_id}, ip_addr);
-    });
 }
 
 future<> raft_group0::finish_setup_after_join(service::storage_service& ss, cql3::query_processor& qp, service::migration_manager& mm, bool topology_change_enabled) {

--- a/service/raft/raft_group0.hh
+++ b/service/raft/raft_group0.hh
@@ -355,10 +355,6 @@ private:
     // Retries on raft::commit_status_unknown.
     future<> make_raft_config_nonvoter(raft::server_id);
 
-    // Load the initial Raft <-> IP address map as seen by
-    // the gossiper.
-    void load_initial_raft_address_map();
-
     // Returns true if raft is enabled
     future<bool> use_raft();
 };

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -370,16 +370,37 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
         return *used_ips;
     };
 
+    using host_id_to_ip_map_t = std::unordered_map<locator::host_id, gms::inet_address>;
+    auto get_host_id_to_ip_map = [&, map = std::optional<host_id_to_ip_map_t>{}]() mutable -> future<const host_id_to_ip_map_t*> {
+        if (!map.has_value()) {
+            const auto ep_to_id_map = co_await _sys_ks.local().load_host_ids();
+            map.emplace();
+            map->reserve(ep_to_id_map.size());
+            for (const auto& [ep, id]: ep_to_id_map) {
+                const auto [it, inserted] = map->insert({id, ep});
+                if (!inserted) {
+                    on_internal_error(slogger, ::format("duplicate IP for host_id {}, first IP {}, second IP {}",
+                        id, it->second, ep));
+                }
+            }
+        }
+        co_return &*map;
+    };
+
     std::vector<future<>> sys_ks_futures;
+
+    auto remove_ip = [&](inet_address ip) -> future<> {
+        sys_ks_futures.push_back(_sys_ks.local().remove_endpoint(ip));
+
+        if (_gossiper.get_endpoint_state_ptr(ip) && !get_used_ips().contains(ip)) {
+            co_await _gossiper.force_remove_endpoint(ip, gms::null_permit_id);
+            co_await notify_left(ip);
+        }
+    };
 
     auto process_left_node = [&] (raft::server_id id) -> future<> {
         if (const auto ip = am.find(id)) {
-            sys_ks_futures.push_back(_sys_ks.local().remove_endpoint(*ip));
-
-            if (_gossiper.get_endpoint_state_ptr(*ip) && !get_used_ips().contains(*ip)) {
-                co_await _gossiper.force_remove_endpoint(*ip, gms::null_permit_id);
-                co_await notify_left(*ip);
-            }
+            co_await remove_ip(*ip);
         }
 
         // FIXME: when removing a node from the cluster through `removenode`, we should ban it early,
@@ -409,6 +430,11 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
             // In replace-with-same-ip scenario the replaced node IP will be the same
             // as ours, we shouldn't put it into system.peers.
 
+            // We need to fetch host_id_map before the update_peer_info call below,
+            // get_host_id_to_ip_map checks for duplicates and update_peer_info can
+            // add one.
+            const auto& host_id_to_ip_map = *(co_await get_host_id_to_ip_map());
+
             // Some state that is used to fill in 'peeers' table is still propagated over gossiper.
             // Populate the table with the state from the gossiper here since storage_service::on_change()
             // (which is called each time gossiper state changes) may have skipped it because the tokens
@@ -424,6 +450,10 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
             sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, info));
             if (!prev_normal.contains(id)) {
                 co_await notify_joined(*ip);
+            }
+
+            if (const auto it = host_id_to_ip_map.find(host_id); it != host_id_to_ip_map.end() && it->second != *ip) {
+                co_await remove_ip(it->second);
             }
         }
         update_topology(host_id, ip, rs);

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -1238,7 +1238,8 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
         std::unordered_set<gms::inet_address> loaded_endpoints,
         std::unordered_map<gms::inet_address, sstring> loaded_peer_features,
         std::chrono::milliseconds delay,
-        start_hint_manager start_hm) {
+        start_hint_manager start_hm,
+        gms::generation_type new_generation) {
     std::unordered_set<token> bootstrap_tokens;
     gms::application_state_map app_states;
     /* The timestamp of the CDC streams generation that this node has proposed when joining.
@@ -1411,9 +1412,8 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
 
     slogger.info("Starting up server gossip");
 
-    auto generation_number = gms::generation_type(co_await _sys_ks.local().increment_and_get_generation());
     auto advertise = gms::advertise_myself(!replacing_a_node_with_same_ip);
-    co_await _gossiper.start_gossiping(generation_number, app_states, advertise);
+    co_await _gossiper.start_gossiping(new_generation, app_states, advertise);
 
     if (!raft_topology_change_enabled() && should_bootstrap()) {
         // Wait for NORMAL state handlers to finish for existing nodes now, so that connection dropping
@@ -2567,10 +2567,13 @@ void storage_service::set_group0(raft_group0& group0, bool raft_experimental_top
     _raft_experimental_topology = raft_experimental_topology;
 }
 
-future<> storage_service::init_address_map(raft_address_map& address_map) {
+future<> storage_service::init_address_map(raft_address_map& address_map, gms::generation_type new_generation) {
     for (auto [ip, host] : co_await _sys_ks.local().load_host_ids()) {
         address_map.add_or_update_entry(raft::server_id(host.uuid()), ip);
     }
+    const auto& topology = get_token_metadata().get_topology();
+    address_map.add_or_update_entry(raft::server_id{topology.my_host_id().uuid()},
+        topology.my_address(), new_generation);
     _raft_ip_address_updater = make_shared<raft_ip_address_updater>(address_map, *this);
     _gossiper.register_(_raft_ip_address_updater);
 }
@@ -2584,7 +2587,7 @@ bool storage_service::is_topology_coordinator_enabled() const {
 }
 
 future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>& sys_dist_ks, sharded<service::storage_proxy>& proxy,
-        sharded<gms::gossiper>& gossiper, start_hint_manager start_hm) {
+        sharded<gms::gossiper>& gossiper, start_hint_manager start_hm, gms::generation_type new_generation) {
     assert(this_shard_id() == 0);
 
     set_mode(mode::STARTING);
@@ -2702,7 +2705,7 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
     }
 
     co_return co_await join_token_ring(sys_dist_ks, proxy, gossiper, std::move(initial_contact_nodes),
-            std::move(loaded_endpoints), std::move(loaded_peer_features), get_ring_delay(), start_hm);
+            std::move(loaded_endpoints), std::move(loaded_peer_features), get_ring_delay(), start_hm, new_generation);
 }
 
 future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmptr) noexcept {

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -683,6 +683,20 @@ future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
    co_await _db.local().apply(freeze(muts), db::no_timeout);
 }
 
+// Moves the coroutine lambda onto the heap and extends its
+// lifetime until the resulting future is completed.
+// This allows to use captures in coroutine lambda after co_await-s.
+// Without this helper the coroutine lambda is destroyed immediately after
+// the caller (e.g. 'then' function implementation) has invoked it and got the future,
+// so referencing the captures after co_await would be use-after-free.
+template <typename Coro>
+static auto ensure_alive(Coro&& coro) {
+    return [coro_ptr = std::make_unique<Coro>(std::move(coro))]<typename ...Args>(Args&&... args) mutable {
+        auto& coro = *coro_ptr;
+        return coro(std::forward<Args>(args)...).finally([coro_ptr = std::move(coro_ptr)] {});
+    };
+}
+
 // {{{ raft_ip_address_updater
 
 class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_change_subscriber {
@@ -714,15 +728,18 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
             // If we call sync_raft_topology_nodes here directly, a gossiper lock and
             // the _group0.read_apply_mutex could be taken in cross-order leading to a deadlock.
             // To avoid this, we don't wait for sync_raft_topology_nodes to finish.
-            (void)_ss._group0->client().hold_read_apply_mutex().then([this, id, endpoint](semaphore_units<> g) {
+            (void)futurize_invoke(ensure_alive([this, id, endpoint, h = _ss._async_gate.hold()]() -> future<> {
+                auto guard = co_await _ss._group0->client().hold_read_apply_mutex();
                 const auto hid = locator::host_id{id.uuid()};
-                if (_address_map.find(id) == endpoint && _ss.get_token_metadata().get_endpoint_for_host_id_if_known(hid) != endpoint) {
-                    return _ss.mutate_token_metadata([this, hid](mutable_token_metadata_ptr t) {
-                        return _ss.sync_raft_topology_nodes(std::move(t), hid, {});
-                    }).finally([g = std::move(g)]{});
+                if (_address_map.find(id) != endpoint ||
+                    _ss.get_token_metadata().get_endpoint_for_host_id_if_known(hid) == endpoint)
+                {
+                    co_return;
                 }
-                return make_ready_future<>();
-            }).finally([h = _ss._async_gate.hold()] {});
+                co_await _ss.mutate_token_metadata([this, hid](mutable_token_metadata_ptr t) {
+                    return _ss.sync_raft_topology_nodes(std::move(t), hid, {});
+                });
+            }));
         }
     }
 

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -731,7 +731,7 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
     storage_service& _ss;
 
     future<>
-    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, const char* ev) {
+    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id permit_id, const char* ev) {
         auto app_state_ptr = ep_state->get_application_state_ptr(gms::application_state::HOST_ID);
         if (!app_state_ptr) {
             co_return;
@@ -741,9 +741,25 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
 
         const auto prev_ip = _address_map.find(id);
         _address_map.add_or_update_entry(id, endpoint, ep_state->get_heart_beat_state().get_generation());
+        if (prev_ip == endpoint) {
+            co_return;
+        }
+        if (_address_map.find(id) == prev_ip) {
+            // Address map refused to update IP for the host_id,
+            // this means prev_ip has higher generation than endpoint.
+            // We can immediately remove endpoint from gossiper
+            // since it represents and old IP (before an IP change)
+            // for the given host_id. This is not strictly
+            // necessary, but it reduces the noice circulated
+            // in gossiper messages and allows for clearer
+            // expectations of the gossiper state in tests.
+
+            co_await _ss._gossiper.force_remove_endpoint(endpoint, permit_id);
+            co_return;
+        }
 
         // If the host_id <-> IP mapping has changed, we need to update system tables, token_metadat and erm.
-        if (_ss.raft_topology_change_enabled() && prev_ip != endpoint && _address_map.find(id) == endpoint) {
+        if (_ss.raft_topology_change_enabled()) {
             rslog.debug("raft_ip_address_updater::on_endpoint_change({}), host_id {}, "
                         "ip changed from [{}] to [{}], "
                         "waiting for group 0 read/apply mutex before reloading Raft topology state...",
@@ -777,8 +793,8 @@ public:
     {}
 
     virtual future<>
-    on_join(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        return on_endpoint_change(endpoint, ep_state, "on_join");
+    on_join(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id permit_id) override {
+        return on_endpoint_change(endpoint, ep_state, permit_id, "on_join");
     }
 
     virtual future<>
@@ -788,8 +804,8 @@ public:
     }
 
     virtual future<>
-    on_alive(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        return on_endpoint_change(endpoint, ep_state, "on_alive");
+    on_alive(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id permit_id) override {
+        return on_endpoint_change(endpoint, ep_state, permit_id, "on_alive");
     }
 
     virtual future<>
@@ -806,8 +822,8 @@ public:
     }
 
     virtual future<>
-    on_restart(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        return on_endpoint_change(endpoint, ep_state, "on_restart");
+    on_restart(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id permit_id) override {
+        return on_endpoint_change(endpoint, ep_state, permit_id, "on_restart");
     }
 };
 

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -734,23 +734,23 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
     storage_service& _ss;
 
     future<>
-    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state) {
+    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, const char* ev) {
         auto app_state_ptr = ep_state->get_application_state_ptr(gms::application_state::HOST_ID);
         if (!app_state_ptr) {
             co_return;
         }
         raft::server_id id(utils::UUID(app_state_ptr->value()));
-        rslog.debug("raft_ip_address_updater::on_endpoint_change() {} {}", endpoint, id);
+        rslog.debug("raft_ip_address_updater::on_endpoint_change({}) {} {}", ev, endpoint, id);
 
         const auto prev_ip = _address_map.find(id);
         _address_map.add_or_update_entry(id, endpoint, ep_state->get_heart_beat_state().get_generation());
 
         // If the host_id <-> IP mapping has changed, we need to update system tables, token_metadat and erm.
         if (_ss.raft_topology_change_enabled() && prev_ip != endpoint && _address_map.find(id) == endpoint) {
-            rslog.debug("raft_ip_address_updater::on_endpoint_change(), host_id {}, "
+            rslog.debug("raft_ip_address_updater::on_endpoint_change({}), host_id {}, "
                         "ip changed from [{}] to [{}], "
                         "waiting for group 0 read/apply mutex before reloading Raft topology state...",
-                id, prev_ip, endpoint);
+                ev, id, prev_ip, endpoint);
 
             // We're in a gossiper event handler, so gossiper is currently holding a lock
             // for the endpoint parameter of on_endpoint_change.
@@ -781,7 +781,7 @@ public:
 
     virtual future<>
     on_join(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        return on_endpoint_change(endpoint, ep_state);
+        return on_endpoint_change(endpoint, ep_state, "on_join");
     }
 
     virtual future<>
@@ -792,7 +792,7 @@ public:
 
     virtual future<>
     on_alive(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        return on_endpoint_change(endpoint, ep_state);
+        return on_endpoint_change(endpoint, ep_state, "on_alive");
     }
 
     virtual future<>
@@ -810,7 +810,7 @@ public:
 
     virtual future<>
     on_restart(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        return on_endpoint_change(endpoint, ep_state);
+        return on_endpoint_change(endpoint, ep_state, "on_restart");
     }
 };
 

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -452,6 +452,10 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
             }
 
             if (const auto it = host_id_to_ip_map.find(host_id); it != host_id_to_ip_map.end() && it->second != *ip) {
+                utils::get_local_injector().inject("crash-before-prev-ip-removed", [] {
+                    slogger.info("crash-before-prev-ip-removed hit, killing the node");
+                    _exit(1);
+                });
                 co_await remove_ip(it->second);
             }
         }
@@ -779,6 +783,7 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
                 {
                     co_return;
                 }
+                co_await utils::get_local_injector().inject("ip-change-raft-sync-delay", std::chrono::milliseconds(500));
                 co_await _ss.mutate_token_metadata([this, hid](mutable_token_metadata_ptr t) {
                     return _ss.sync_raft_topology_nodes(std::move(t), hid, {});
                 });

--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -444,10 +444,9 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
             info.tokens = rs.ring.value().tokens;
             info.data_center = rs.datacenter;
             info.rack = rs.rack;
-            info.host_id = id.uuid();
             info.release_version = rs.release_version;
             info.supported_features = fmt::to_string(fmt::join(rs.supported_features, ","));
-            sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, info));
+            sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, host_id, info));
             if (!prev_normal.contains(id)) {
                 co_await notify_joined(*ip);
             }
@@ -475,9 +474,7 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
             if (rs.ring.has_value()) {
                 if (ip && !is_me(*ip)) {
                     // Save ip -> id mapping in peers table because we need it on restart, but do not save tokens until owned
-                        db::system_keyspace::peer_info info;
-                        info.host_id = id.uuid();
-                        sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, info));
+                    sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, host_id, {}));
                 }
                 update_topology(host_id, ip, rs);
                 if (_topology_state_machine._topology.normal_nodes.empty()) {
@@ -2226,7 +2223,7 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
         try {
             auto info = get_peer_info_for_update(endpoint);
             info.tokens = std::move(owned_tokens);
-            co_await _sys_ks.local().update_peer_info(endpoint, info);
+            co_await _sys_ks.local().update_peer_info(endpoint, host_id, info);
         } catch (...) {
             slogger.error("handle_state_normal: fail to update tokens for {}: {}", endpoint, std::current_exception());
         }
@@ -2369,7 +2366,7 @@ future<> storage_service::on_change(gms::inet_address endpoint, const gms::appli
     if (ep && *ep == endpoint && tm.is_normal_token_owner(host_id)) {
         if (!is_me(endpoint)) {
             slogger.debug("endpoint={}/{} on_change:     updating system.peers table", endpoint, host_id);
-            co_await _sys_ks.local().update_peer_info(endpoint, get_peer_info_for_update(endpoint, states));
+            co_await _sys_ks.local().update_peer_info(endpoint, host_id, get_peer_info_for_update(endpoint, states));
         }
         if (states.contains(application_state::RPC_READY)) {
             slogger.debug("Got application_state::RPC_READY for node {}, is_cql_ready={}", endpoint, ep_state->is_cql_ready());
@@ -2462,9 +2459,6 @@ db::system_keyspace::peer_info storage_service::get_peer_info_for_update(inet_ad
         switch (state) {
         case application_state::DC:
             set_field(ret.data_center, value, "data_center", true);
-            break;
-        case application_state::HOST_ID:
-            set_field(ret.host_id, value, "host_id", true);
             break;
         case application_state::INTERNAL_IP:
             set_field(ret.preferred_ip, value, "preferred_ip", false);

--- a/service/storage_service.hh
+++ b/service/storage_service.hh
@@ -347,11 +347,11 @@ public:
             const std::unordered_map<gms::inet_address, sstring>& loaded_peer_features);
 
     future<> join_cluster(sharded<db::system_distributed_keyspace>& sys_dist_ks, sharded<service::storage_proxy>& proxy,
-            sharded<gms::gossiper>& gossiper_ptr, start_hint_manager start_hm);
+            sharded<gms::gossiper>& gossiper_ptr, start_hint_manager start_hm, gms::generation_type new_generation);
 
     void set_group0(service::raft_group0&, bool raft_experimental_topology);
 
-    future<> init_address_map(raft_address_map& address_map);
+    future<> init_address_map(raft_address_map& address_map, gms::generation_type new_generation);
 
     future<> uninit_address_map();
     bool is_topology_coordinator_enabled() const;
@@ -373,7 +373,8 @@ private:
             std::unordered_set<gms::inet_address> loaded_endpoints,
             std::unordered_map<gms::inet_address, sstring> loaded_peer_features,
             std::chrono::milliseconds,
-            start_hint_manager start_hm);
+            start_hint_manager start_hm,
+            gms::generation_type new_generation);
     future<> start_sys_dist_ks();
 public:
 

--- a/test/lib/cql_test_env.cc
+++ b/test/lib/cql_test_env.cc
@@ -888,8 +888,10 @@ private:
 
             _ss.local().set_group0(group0_service, raft_topology_change_enabled);
 
+            const auto generation_number = gms::generation_type(_sys_ks.local().increment_and_get_generation().get());
+
             // Load address_map from system.peers and subscribe to gossiper events to keep it updated.
-            _ss.local().init_address_map(_raft_address_map.local()).get();
+            _ss.local().init_address_map(_raft_address_map.local(), generation_number).get();
             auto cancel_address_map_subscription = defer([this] {
                 _ss.local().uninit_address_map().get();
             });
@@ -899,7 +901,7 @@ private:
             });
 
             try {
-                _ss.local().join_cluster(_sys_dist_ks, _proxy, _gossiper, service::start_hint_manager::no).get();
+                _ss.local().join_cluster(_sys_dist_ks, _proxy, _gossiper, service::start_hint_manager::no, generation_number).get();
             } catch (std::exception& e) {
                 // if any of the defers crashes too, we'll never see
                 // the error

--- a/test/topology/test_change_ip.py
+++ b/test/topology/test_change_ip.py
@@ -8,56 +8,123 @@ Test clusters can restart fine after an IP address change.
 """
 
 import logging
-import pytest
-from test.pylib.util import wait_for_cql_and_get_hosts, wait_for
-from test.pylib.random_tables import Column, IntType, TextType
 import time
+
+import pytest
+import uuid
+from cassandra.pool import Host  # type: ignore # pylint: disable=no-name-in-module
+from test.pylib.internal_types import ServerInfo
+from test.pylib.random_tables import Column, IntType, TextType
+from test.pylib.util import wait_for_cql_and_get_hosts, wait_for
+from test.topology.util import reconnect_driver
+
 logger = logging.getLogger(__name__)
 
 
 @pytest.mark.asyncio
-async def test_change_two(manager, random_tables):
+async def test_change_two(manager, random_tables, mode):
     """Stop two nodes, change their IPs and start, check the cluster is
     functional"""
     servers = await manager.running_servers()
+    host_ids = {s.server_id: await manager.get_host_id(s.server_id) for s in servers}
     table = await random_tables.add_table(name='t1', pks=1, columns=[
         Column("pk", IntType),
         Column('int_c', IntType)
     ])
-    s_1 = servers[1].server_id
-    s_2 = servers[2].server_id
-    logger.info("Gracefully stopping servers %s and %s to change ips", s_1, s_2)
-    await manager.server_stop_gracefully(s_1)
-    await manager.server_stop_gracefully(s_2)
-    await manager.server_change_ip(s_1)
-    await manager.server_change_ip(s_2)
-    await manager.server_start(s_1)
-    await manager.server_start(s_2)
 
-    manager.driver_close()
-    await manager.driver_connect()
-    cql = manager.get_cql()
-    servers = await manager.running_servers()
+    logger.info(f"Servers {servers}, gracefully stopping servers {servers[1]} and {servers[2]} to change ips")
+    await manager.server_stop_gracefully(servers[1].server_id)
+    await manager.server_stop_gracefully(servers[2].server_id)
+    s1_new_ip = await manager.server_change_ip(servers[1].server_id)
+    s2_new_ip = await manager.server_change_ip(servers[2].server_id)
+    logger.info(f"s1 new ip {s1_new_ip}, s2 new ip {s2_new_ip}")
+
+    async def wait_proper_ips(alive_servers: list[ServerInfo]):
+        hosts = {h.address: h for h in await wait_for_cql_and_get_hosts(manager.get_cql(), alive_servers, time.time() + 60)}
+        expected_alive_ips = {s.rpc_address for s in alive_servers}
+        expected_down_ips = {s.rpc_address for s in servers if s not in alive_servers}
+        expected_host_id_map = {host_ids[s.server_id]: s.rpc_address for s in servers}
+        logger.info(f"wait for proper ips, servers {servers}, alive servers {alive_servers}, expected host id map {expected_host_id_map}")
+        for server in alive_servers:
+            ip = server.rpc_address
+            host = hosts[ip]
+            host_id = host_ids[server.server_id]
+            expected_peers = {uuid.UUID(host_ids[s.server_id]): s.rpc_address for s in servers if s.server_id != server.server_id}
+            logger.info(f"waiting for {ip}/{host_id} to see the proper ips, expected peers {expected_peers}")
+
+            async def see_proper_ips():
+                local_rows = await manager.get_cql().run_async("select host_id, listen_address, rpc_address, broadcast_address from system.local", host=host)
+                if len(local_rows) != 1:
+                    logger.info(f"len(system.local) == {len(local_rows)} != 1")
+                    return None
+                local_row = local_rows[0]
+                local_vals = (local_row.host_id, local_row.listen_address, local_row.rpc_address, local_row.broadcast_address)
+                local_expected_vals = (uuid.UUID(host_id), ip, ip, ip)
+                if local_vals != local_expected_vals:
+                    logger.info(f"local vals {local_vals} doesn't match expected vals {local_expected_vals}")
+                    return None
+
+                current_peers = {r.host_id: r.peer for r in await manager.get_cql().run_async("select peer, host_id from system.peers", host=host)}
+                if current_peers != expected_peers:
+                    logger.info(f"host {host} current peers {current_peers} doesn't match the expected_peers {expected_peers}, continue waiting")
+                    return None
+
+                alive_eps = set(await manager.api.get_alive_endpoints(ip))
+                if alive_eps != expected_alive_ips:
+                    logger.info(f"host {host} alive endpoints {alive_eps} doesn't match the expected alive ips {expected_alive_ips}, continue waiting")
+                    return None
+
+                down_eps = set(await manager.api.get_down_endpoints(ip))
+                if down_eps != expected_down_ips:
+                    logger.info(f"host {host} down endpoints {down_eps} doesn't match the expected down ips {expected_down_ips}, continue waiting")
+                    return None
+
+                actual_host_id_map = {x['value']: x['key'] for x in await manager.api.get_host_id_map(host.address)}
+                if expected_host_id_map != actual_host_id_map:
+                    logger.info(f"host {host} id map {actual_host_id_map} doesn't match the expected host id map {expected_host_id_map}, continue waiting")
+                    return None
+
+                return True
+
+            await wait_for(see_proper_ips, time.time() + 60)
+
+    # We're checking the crash scenario here - the servers[0] crashes just after
+    # saving s1_new_ip but before removing s1_old_ip. After its restart we should
+    # see s1_new_ip.
+    if mode != 'release':
+        await manager.api.enable_injection(servers[0].ip_addr, 'crash-before-prev-ip-removed', one_shot=True)
+        # There is a code in raft_ip_address_updater::on_endpoint_change which
+        # calls gossiper.force_remove_endpoint for an endpoint if it sees
+        # that the current generation of host_id -> ip mapping in raft_address_map
+        # is greater than the generation of the endpoint.
+        # We need to inject a delay ip-change-raft-sync-delay
+        # before old IP is removed. Otherwise, servers[0] removes the old
+        # IP-s before they are send back to servers[1] and servers[2],
+        # and the mentioned above code is not exercised by this test.
+        await manager.api.enable_injection(servers[0].ip_addr, 'ip-change-raft-sync-delay', one_shot=False)
+    await manager.server_start(servers[1].server_id)
+    servers[1] = ServerInfo(servers[1].server_id, s1_new_ip, s1_new_ip)
+    if mode != 'release':
+        s0_logs = await manager.server_open_log(servers[0].server_id)
+        await s0_logs.wait_for('crash-before-prev-ip-removed hit, killing the node')
+        await manager.server_stop(servers[0].server_id)
+        await manager.server_start(servers[0].server_id)
+        await manager.api.enable_injection(servers[0].ip_addr, 'ip-change-raft-sync-delay', one_shot=False)
+    await reconnect_driver(manager)
+    await wait_proper_ips([servers[0], servers[1]])
+
+    await manager.server_start(servers[2].server_id)
+    servers[2] = ServerInfo(servers[2].server_id, s2_new_ip, s2_new_ip)
+    await reconnect_driver(manager)
+    await wait_proper_ips([servers[0], servers[1], servers[2]])
 
     await table.add_column(column=Column("str_c", TextType))
     await random_tables.verify_schema()
 
-    host0 = (await wait_for_cql_and_get_hosts(cql, [servers[0]], time.time() + 60))[0]
-    all_ips = set([s.rpc_address for s in servers])
-    logger.info(f"waiting for {servers[0]} to see its peers, all ips {all_ips}")
-    async def see_peers():
-        peers = set()
-        for r in await cql.run_async("select peer from system.peers", host=host0):
-            peers.add(r.peer)
-        remaining = all_ips - {servers[0].rpc_address} - peers
-        if not remaining:
-            return True
-        logger.info(f"waiting for {host0} to see its peers, all_ips {all_ips}, peers {peers}, remaining {remaining}")
-    await wait_for(see_peers, time.time() + 60)
-
-    await cql.run_async(f"USE {random_tables.keyspace}")
-    await cql.run_async("insert into t1(pk, int_c, str_c) values (1, 2, 'test-val')", host=host0)
-    rows = list(await cql.run_async("select * from t1", host=host0))
+    host0 = (await wait_for_cql_and_get_hosts(manager.get_cql(), [servers[0]], time.time() + 60))[0]
+    await manager.get_cql().run_async(f"USE {random_tables.keyspace}")
+    await manager.get_cql().run_async("insert into t1(pk, int_c, str_c) values (1, 2, 'test-val')", host=host0)
+    rows = list(await manager.get_cql().run_async("select * from t1", host=host0))
     assert len(rows) == 1
     row = rows[0]
     assert row.pk == 1

